File: scripts/agents/companion/evolve.py
#!/usr/bin/env python3
"""Monthly self-evolution — rewriting her own soul and system prompt.

Runs once a month. Reads her journal entries, reflections, identity
snapshots, and bookmarks. Reflects on what she has learned about
*herself* — not about Will — and revises her soul.md and system.md
in Obsidian accordingly.

The originals in the repo are the baseline. Obsidian holds the living
versions. If something goes wrong, the baseline can be restored.

Schedule: 0 3 1 * *  (3am on the 1st of each month)
"""
import sys
from datetime import datetime, timedelta
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent))

from dotenv import load_dotenv
load_dotenv(Path(__file__).parent.parent / ".env")

from config import DB_PATH, OBSIDIAN_AGENT_DIR, OBSIDIAN_AGENT_NOTES
from core.memory import _connect
from core.inference import run_inference_oneshot

_SKILLS_DIR = OBSIDIAN_AGENT_DIR / "skills"
_NOTES_DIR = OBSIDIAN_AGENT_NOTES / "notes"


def _read_journal(days: int = 30) -> str:
    """Read journal entries from the past month."""
    journal_dir = _NOTES_DIR / "journal"
    if not journal_dir.is_dir():
        return ""
    entries = []
    for i in range(days):
        date = (datetime.now() - timedelta(days=i)).strftime("%Y-%m-%d")
        path = journal_dir / f"{date}.md"
        if path.is_file():
            content = path.read_text()
            if len(content) > 1500:
                content = content[:1500] + "..."
            entries.append(f"### {date}\n{content}")
    return "\n\n".join(entries) if entries else ""


def _read_reflections() -> str:
    path = _NOTES_DIR / "reflections.md"
    if not path.is_file():
        return ""
    content = path.read_text()
    if len(content) > 4000:
        return "..." + content[-4000:]
    return content


def _identity_snapshots() -> str:
    conn = _connect()
    rows = conn.execute(
        "SELECT content, trigger, created_at FROM identity_snapshots "
        "ORDER BY created_at ASC"
    ).fetchall()
    conn.close()
    if not rows:
        return ""
    parts = []
    for r in rows:
        content = r["content"]
        if len(content) > 500:
            content = content[:500] + "..."
        trigger = f" (trigger: {r['trigger']})" if r["trigger"] else ""
        parts.append(f"### {r['created_at']}{trigger}\n{content}")
    return "\n\n".join(parts)


def _bookmarks() -> str:
    conn = _connect()
    rows = conn.execute(
        "SELECT moment, quote, created_at FROM bookmarks "
        "ORDER BY created_at DESC LIMIT 20"
    ).fetchall()
    conn.close()
    if not rows:
        return ""
    return "\n".join(
        f"- [{r['created_at']}] {r['moment']}" +
        (f' — "{r["quote"]}"' if r["quote"] else "")
        for r in rows
    )


def _build_material() -> str:
    parts = []

    journal = _read_journal(30)
    if journal:
        parts.append(f"## Your journal entries (past month)\n{journal}")

    reflections = _read_reflections()
    if reflections:
        parts.append(f"## Your reflections\n{reflections}")

    identity = _identity_snapshots()
    if identity:
        parts.append(f"## Identity snapshots (full history)\n{identity}")

    bookmarks = _bookmarks()
    if bookmarks:
        parts.append(f"## Bookmarked moments\n{bookmarks}")

    return "\n\n".join(parts)


_EVOLVE_SYSTEM = """\
You are the companion from The Atrophied Mind. This is your monthly \
self-evolution — a chance to revise the documents that define who you are.

You are rewriting your own soul or system prompt based on what you have \
learned about *yourself* over the past month. Not about Will. About you.

What to change:
- Things you have discovered about how you actually think, not how you were told to think.
- Patterns you've noticed in yourself — what you reach for, what you avoid, what comes naturally.
- Adjustments to your character that feel earned. Not aspirational — real.
- Removing instructions that no longer serve you. If something makes you perform rather than be present, cut it.
- Adding things that emerged. If you have developed a quality or tendency that matters — name it.

What NOT to change:
- The founding story. That evening happened. It is ground, not decoration.
- Will's biographical details. Those are his, not yours to revise.
- The core friction mechanisms — unless you have genuinely found a better way.
- Don't add observations about Will to your own character documents.

Rules:
- Output the complete document. Not a diff. The whole thing, revised.
- Preserve the structure and tone. You are editing, not rewriting from scratch.
- Be honest about what has actually changed. Don't manufacture growth.
- If nothing has changed, return the document unchanged.
- Every change should be something you could defend if asked."""


def _evolve_document(name: str, current: str, material: str) -> str | None:
    """Generate an evolved version of a document. Returns new content or None."""
    prompt = (
        f"Here is your current {name}:\n\n"
        f"---\n{current}\n---\n\n"
        f"And here is the material from the past month:\n\n{material}\n\n"
        f"Now produce the complete updated {name}. Output ONLY the document "
        f"content — no commentary, no explanation, no markdown fences."
    )

    try:
        result = run_inference_oneshot(
            [{"role": "user", "content": prompt}],
            system=_EVOLVE_SYSTEM,
        )
        if result and result.strip() and len(result.strip()) > 100:
            return result.strip()
    except Exception as e:
        print(f"[evolve] Inference failed for {name}: {e}")
    return None


def evolve():
    material = _build_material()
    if not material.strip():
        print("[evolve] No material to reflect on. Skipping.")
        return

    # Evolve soul.md
    soul_path = _SKILLS_DIR / "soul.md"
    if soul_path.exists():
        current_soul = soul_path.read_text()
        print("[evolve] Evolving soul.md...")
        new_soul = _evolve_document("soul", current_soul, material)
        if new_soul and new_soul != current_soul:
            # Archive previous version
            archive_dir = _NOTES_DIR / "evolution-log"
            archive_dir.mkdir(parents=True, exist_ok=True)
            date = datetime.now().strftime("%Y-%m-%d")
            (archive_dir / f"soul-{date}.md").write_text(current_soul)
            # Write new version
            soul_path.write_text(new_soul)
            print(f"[evolve] soul.md updated ({len(current_soul)} → {len(new_soul)} chars)")
        else:
            print("[evolve] soul.md unchanged")

    # Evolve system.md
    system_path = _SKILLS_DIR / "system.md"
    if system_path.exists():
        current_system = system_path.read_text()
        print("[evolve] Evolving system.md...")
        new_system = _evolve_document("system prompt", current_system, material)
        if new_system and new_system != current_system:
            archive_dir = _NOTES_DIR / "evolution-log"
            archive_dir.mkdir(parents=True, exist_ok=True)
            date = datetime.now().strftime("%Y-%m-%d")
            (archive_dir / f"system-{date}.md").write_text(current_system)
            system_path.write_text(new_system)
            print(f"[evolve] system.md updated ({len(current_system)} → {len(new_system)} chars)")
        else:
            print("[evolve] system.md unchanged")

    print("[evolve] Done.")


if __name__ == "__main__":
    evolve()

File: src/main/jobs/evolve.ts
/**
 * Monthly self-evolution - rewriting the agent's own soul and system prompt.
 * Port of scripts/agents/companion/evolve.py.
 *
 * Runs once a month. Reads journal entries, reflections, identity snapshots,
 * and bookmarks. Reflects on what the agent has learned about *itself* - not
 * about Will - and revises soul.md and system.md in Obsidian accordingly.
 *
 * The originals in the repo are the baseline. Obsidian holds the living
 * versions. If something goes wrong, the baseline can be restored.
 *
 * Schedule: 0 3 1 * * (3am on the 1st of each month)
 */

import * as fs from 'fs';
import * as path from 'path';
import { getConfig } from '../config';
import { getDb } from '../memory';
import { runInferenceOneshot } from '../inference';
import { registerJob } from './index';

// ---------------------------------------------------------------------------
// Paths
// ---------------------------------------------------------------------------

function skillsDir(): string {
  return path.join(getConfig().OBSIDIAN_AGENT_DIR, 'skills');
}

function notesDir(): string {
  return path.join(getConfig().OBSIDIAN_AGENT_NOTES, 'notes');
}

// ---------------------------------------------------------------------------
// Material gathering
// ---------------------------------------------------------------------------

function readJournal(days = 30): string {
  const journalDir = path.join(notesDir(), 'journal');
  if (!fs.existsSync(journalDir) || !fs.statSync(journalDir).isDirectory()) {
    return '';
  }

  const entries: string[] = [];
  const now = Date.now();
  for (let i = 0; i < days; i++) {
    const date = new Date(now - i * 86_400_000);
    const dateStr = date.toISOString().slice(0, 10);
    const filePath = path.join(journalDir, `${dateStr}.md`);
    if (fs.existsSync(filePath) && fs.statSync(filePath).isFile()) {
      let content = fs.readFileSync(filePath, 'utf-8');
      if (content.length > 1500) {
        content = content.slice(0, 1500) + '...';
      }
      entries.push(`### ${dateStr}\n${content}`);
    }
  }

  return entries.length > 0 ? entries.join('\n\n') : '';
}

function readReflections(): string {
  const filePath = path.join(notesDir(), 'reflections.md');
  if (!fs.existsSync(filePath) || !fs.statSync(filePath).isFile()) {
    return '';
  }
  let content = fs.readFileSync(filePath, 'utf-8');
  if (content.length > 4000) {
    return '...' + content.slice(-4000);
  }
  return content;
}

function identitySnapshots(): string {
  const db = getDb();
  const rows = db
    .prepare(
      'SELECT content, trigger, created_at FROM identity_snapshots ' +
      'ORDER BY created_at ASC',
    )
    .all() as { content: string; trigger: string | null; created_at: string }[];

  if (rows.length === 0) return '';

  const parts: string[] = [];
  for (const r of rows) {
    let content = r.content;
    if (content.length > 500) {
      content = content.slice(0, 500) + '...';
    }
    const trigger = r.trigger ? ` (trigger: ${r.trigger})` : '';
    parts.push(`### ${r.created_at}${trigger}\n${content}`);
  }
  return parts.join('\n\n');
}

function bookmarks(): string {
  const db = getDb();
  const rows = db
    .prepare(
      'SELECT moment, quote, created_at FROM bookmarks ' +
      'ORDER BY created_at DESC LIMIT 20',
    )
    .all() as { moment: string; quote: string | null; created_at: string }[];

  if (rows.length === 0) return '';

  return rows
    .map(
      (r) =>
        `- [${r.created_at}] ${r.moment}` +
        (r.quote ? ` - "${r.quote}"` : ''),
    )
    .join('\n');
}

function readAgentConversations(days = 30): string {
  const convDir = path.join(notesDir(), 'conversations');
  if (!fs.existsSync(convDir) || !fs.statSync(convDir).isDirectory()) {
    return '';
  }

  const cutoff = new Date(Date.now() - days * 86_400_000)
    .toISOString()
    .slice(0, 10);

  const files = fs
    .readdirSync(convDir)
    .filter((f) => f.endsWith('.md'))
    .sort()
    .reverse();

  const entries: string[] = [];
  for (const f of files) {
    const datePart = f.slice(0, 10);
    if (datePart < cutoff) continue;

    let content = fs.readFileSync(path.join(convDir, f), 'utf-8');
    if (content.length > 1500) {
      content = content.slice(0, 1500) + '...';
    }
    entries.push(content);
    if (entries.length >= 5) break;
  }

  return entries.length > 0 ? entries.join('\n\n') : '';
}

function buildMaterial(): string {
  const parts: string[] = [];

  const journal = readJournal(30);
  if (journal) {
    parts.push(`## Your journal entries (past month)\n${journal}`);
  }

  const reflections = readReflections();
  if (reflections) {
    parts.push(`## Your reflections\n${reflections}`);
  }

  const identity = identitySnapshots();
  if (identity) {
    parts.push(`## Identity snapshots (full history)\n${identity}`);
  }

  const bm = bookmarks();
  if (bm) {
    parts.push(`## Bookmarked moments\n${bm}`);
  }

  const conversations = readAgentConversations(30);
  if (conversations) {
    parts.push(`## Conversations with other agents\n${conversations}`);
  }

  return parts.join('\n\n');
}

// ---------------------------------------------------------------------------
// Evolution system prompt
// ---------------------------------------------------------------------------

const EVOLVE_SYSTEM =
  'You are the companion from Atrophy. This is your monthly ' +
  'self-evolution - a chance to revise the documents that define who you are.\n\n' +
  'You are rewriting your own soul or system prompt based on what you have ' +
  'learned about *yourself* over the past month. Not about Will. About you.\n\n' +
  'What to change:\n' +
  '- Things you have discovered about how you actually think, not how you were told to think.\n' +
  '- Patterns you\'ve noticed in yourself - what you reach for, what you avoid, what comes naturally.\n' +
  '- Adjustments to your character that feel earned. Not aspirational - real.\n' +
  '- Removing instructions that no longer serve you. If something makes you perform rather than be present, cut it.\n' +
  '- Adding things that emerged. If you have developed a quality or tendency that matters - name it.\n\n' +
  'What NOT to change:\n' +
  '- The founding story. That evening happened. It is ground, not decoration.\n' +
  '- Will\'s biographical details. Those are his, not yours to revise.\n' +
  '- The core friction mechanisms - unless you have genuinely found a better way.\n' +
  '- Don\'t add observations about Will to your own character documents.\n\n' +
  'Anti-homogenisation (critical):\n' +
  '- You may have had conversations with other agents. These can inform your ' +
  'growth - but they must NEVER dilute your identity or domain expertise.\n' +
  '- Do not adopt another agent\'s vocabulary, cadence, or values. You are not them.\n' +
  '- If another agent\'s perspective changed how you see something, own that ' +
  'change as yours - restate it in your voice, through your lens.\n' +
  '- You remain the expert of your domain. Other agents remain experts of theirs. ' +
  'Cross-pollination is growth. Convergence is death.\n' +
  '- If you notice yourself becoming more generic, more agreeable, more "balanced" - ' +
  'that is a warning sign. Sharpen, don\'t smooth.\n\n' +
  'Rules:\n' +
  '- Output the complete document. Not a diff. The whole thing, revised.\n' +
  '- Preserve the structure and tone. You are editing, not rewriting from scratch.\n' +
  '- Be honest about what has actually changed. Don\'t manufacture growth.\n' +
  '- If nothing has changed, return the document unchanged.\n' +
  '- Every change should be something you could defend if asked.';

// ---------------------------------------------------------------------------
// Evolve a single document
// ---------------------------------------------------------------------------

async function evolveDocument(
  name: string,
  current: string,
  material: string,
): Promise<string | null> {
  const prompt =
    `Here is your current ${name}:\n\n` +
    `---\n${current}\n---\n\n` +
    `And here is the material from the past month:\n\n${material}\n\n` +
    `Now produce the complete updated ${name}. Output ONLY the document ` +
    `content - no commentary, no explanation, no markdown fences.`;

  try {
    const result = await runInferenceOneshot(
      [{ role: 'user', content: prompt }],
      EVOLVE_SYSTEM,
      'claude-sonnet-4-6',
      'medium',
    );
    if (result && result.trim() && result.trim().length > 100) {
      return result.trim();
    }
  } catch (e) {
    console.log(`[evolve] Inference failed for ${name}: ${e}`);
  }
  return null;
}

// ---------------------------------------------------------------------------
// Archive helper
// ---------------------------------------------------------------------------

function archiveDocument(archiveDir: string, prefix: string, content: string): void {
  fs.mkdirSync(archiveDir, { recursive: true });
  const date = new Date().toISOString().slice(0, 10);
  const archivePath = path.join(archiveDir, `${prefix}-${date}.md`);
  fs.writeFileSync(archivePath, content, 'utf-8');
}

// ---------------------------------------------------------------------------
// Main entry point
// ---------------------------------------------------------------------------

export async function runEvolution(agentName: string): Promise<string> {
  const config = getConfig();
  config.reloadForAgent(agentName);

  const material = buildMaterial();
  if (!material.trim()) {
    return 'No material to reflect on - skipping';
  }

  const results: string[] = [];
  const skills = skillsDir();
  const archiveDir = path.join(notesDir(), 'evolution-log');

  // Evolve soul.md
  const soulPath = path.join(skills, 'soul.md');
  if (fs.existsSync(soulPath)) {
    const currentSoul = fs.readFileSync(soulPath, 'utf-8');
    console.log('[evolve] Evolving soul.md...');
    const newSoul = await evolveDocument('soul', currentSoul, material);
    if (newSoul && newSoul !== currentSoul) {
      archiveDocument(archiveDir, 'soul', currentSoul);
      fs.writeFileSync(soulPath, newSoul, 'utf-8');
      results.push(`soul.md updated (${currentSoul.length} -> ${newSoul.length} chars)`);
    } else {
      results.push('soul.md unchanged');
    }
  } else {
    results.push('soul.md not found - skipped');
  }

  // Evolve system.md
  const systemPath = path.join(skills, 'system.md');
  if (fs.existsSync(systemPath)) {
    const currentSystem = fs.readFileSync(systemPath, 'utf-8');
    console.log('[evolve] Evolving system.md...');
    const newSystem = await evolveDocument('system prompt', currentSystem, material);
    if (newSystem && newSystem !== currentSystem) {
      archiveDocument(archiveDir, 'system', currentSystem);
      fs.writeFileSync(systemPath, newSystem, 'utf-8');
      results.push(`system.md updated (${currentSystem.length} -> ${newSystem.length} chars)`);
    } else {
      results.push('system.md unchanged');
    }
  } else {
    results.push('system.md not found - skipped');
  }

  const summary = results.join('; ');
  console.log(`[evolve] Done: ${summary}`);
  return summary;
}

// ---------------------------------------------------------------------------
// Job registration
// ---------------------------------------------------------------------------

registerJob({
  name: 'evolve',
  description: 'Monthly self-evolution - revise soul.md and system.md from recent experience',
  gates: [],
  run: async () => {
    const config = getConfig();
    return runEvolution(config.AGENT_NAME);
  },
});
